stdlib/public/SwiftShims/FoundationOverlayShims.h

```diff
@@ -29,3 +32,41 @@
 #import "NSLocaleShims.h"
 #import "NSTimeZoneShims.h"
 #import "NSUndoManagerShims.h"
+
+typedef struct {
+    void *_Nonnull memory;
+    size_t capacity;
+    _Bool onStack;
+} _ConditionalAllocationBuffer;
+
+static inline _Bool _resizeConditionalAllocationBuffer(_ConditionalAllocationBuffer *_Nonnull buffer, size_t amt) {
+    size_t amount = malloc_good_size(amt);
+    if (amount <= buffer->capacity) { return true; }
+    void *newMemory;
+    if (buffer->onStack) {
+        newMemory = malloc(amount);
+        if (newMemory == NULL) { return false; }
+        memcpy(newMemory, buffer->memory, buffer->capacity);
+        buffer->onStack = false;
+    } else {
+        newMemory = realloc(buffer->memory, amount);
+        if (newMemory == NULL) { return false; }
+    }
+    if (newMemory == NULL) { return false; }
+    buffer->memory = newMemory;
+    buffer->capacity = amount;
+    return true;
+}
+
+static inline _Bool _withStackOrHeapBuffer(size_t amount, void (__attribute__((noescape)) ^ _Nonnull applier)(_ConditionalAllocationBuffer *_Nonnull)) {
+    _ConditionalAllocationBuffer buffer;
+    buffer.capacity = malloc_good_size(amount);
```

Comment:
Why round up prior to determining `onStack`? If the requested amount of memory is 500 bytes, `malloc_good_size` will probably return 512 or higher, which puts it out of stack viability (on the main thread). From a quick reading, it seems like we intend for allocations of that size to be on the stack.

I'm curious what the benefit to `malloc_good_size` is here at all. I suppose if you expected it to grow, you'd want to `malloc_good_size` to give you some more time before needing to realloc, but I don't actually see `_resizeConditionalAllocationBuffer` being called anywhere.